### PR TITLE
Including all SFA Functions

### LINKED ISSUE #11: Include all functions already described in Simple Features for SQL
**Official change request**: [OGC CR 599](Include all functions already described in Simple Features for SQL)

GeoSPARQL describes a number of SPARQL functions that can be implemented by developers of SPARQL endpoints. It would be beneficial for working with spatial data in SPARQL if more ways of interacting with spatial data were available. The routines and functions described in the Simple Features for SQL specification (06-104r4) are well known and have been implemented in different libraries and data storage systems. It could be a relatively simple effort to transpose the existing definitions to RDFS/OWL, so developers of SPARQL systems can implement those often very useful functions, either by using existing libraries, or by new development.

### LINKED ISSUE #160: Function to retrieve lat/lon
I just stumbled upon this StackOverflow post from many years ago asking how to retrieve lat/long from a GeoSPARQL-encoded geometry.
https://stackoverflow.com/questions/22536467/how-to-retrieve-latitude-and-longitude-from-geosparqls-wktliteral
The solution seems rather cumbersome.
Now with GeoSPARQL 1.1 we could (mis-)use e.g. the maxX() and maxY() function to retrieve lat/long from a single point, but it really seems like a fundamental thing to do that is currently not well supported in GeoSPARQL.
Any opinions about this?

### PR DESCRIPTION
**Pull request by situx**
Beginning to add more SFA function definitions for GeoSPARQL 1.2
Closes #160
Closes #11

### CODE REVIEW

**Comment by situx** [comment]
As far as the Simple Features functions are concerned, all functions have been added to this pull request. What remains to be done is to discuss the issues I pointed out in the first post, to resolve them and to review the work done.

File: 1.1/funcsrules.ttl
Comment (mperry455) [inline 1.1/funcsrules.ttl:681]: typo: "equivalent WKT representation" --> "equivalent WKB representation"

File: 1.1/geo.ttl
Comment (mperry455) [inline 1.1/geo.ttl:809]: Do you mean isClosed?

File: 1.1/spec/11-Part-08.adoc
Comment (mperry455) [inline 1.1/spec/11-Part-08.adoc:1402]: Do we need this one in funcsrules.ttl?

File: 1.1/spec/11-Part-08.adoc
Comment (mperry455) [inline 1.1/spec/11-Part-08.adoc:1063]: Multipolygon --> PolyhedralSurface?

**Comment by situx** [comment]
There is another issue we need to consider here:
https://libgeos.org/specifications/wkb/
As you can see, WKB is not sufficient to encode 3D geometries.
To allow for the encoding of 3D geometries we need to allow EWKB (Extended WKB), so that is what I will propose in this pull request instead.

**Comment by nicholascar** [comment]
I don't have experience with the use of WKB, TinyWKB or EWKB, or experience with the community that uses them, to know of the pros and cons here.

My quick reading of the format descriptors seems to indicate TinyWKB is a lot better than WKB but that's just from reading the spec, not use.

Since we don't have a legacy of supporting any WKB format so far, and we are very interested in 3D support in this standard version, I suggest we go for the formats that support 3D, so this is TinyWKB & EWKB.

For WKT, we actually already support a form of EWTK now, since we specify how to include the SRID.

Perhaps we should clarify a policy on literal definitions: do we try and define as many as possible, like 3 x WKB, 2 x WKT, and leave it up to implementers to actually realise or do we try to pick some winners like perhaps only EWKB & EWKT, stealing from Postgres?

**Comment by situx** [comment]
In today's meeting we thought that the definition of EKWB would be enough since it covers 3D and all WKB definitions. TWKB did not seem to be widely used.
I will adjust the specification accordingly so we only define one EWKB literal with Base64 encoding with a URI definition preceding the Base64 encoded EWKB String.
With that the pull request would again be open for review and approvals/rejects. 
It will be merged once we have established in which version/branch of the document to merge it.

File: 1.1/funcsrules.ttl
Comment (situx) [reply 1.1/funcsrules.ttl]: Fixed

File: 1.1/geo.ttl
Comment (situx) [reply 1.1/geo.ttl]: Yes. It is fixed now

File: 1.1/spec/11-Part-08.adoc
Comment (situx) [reply 1.1/spec/11-Part-08.adoc]: Yes, that fits better. Fixed now

**Comment by situx** [comment]
I updated the pull request yet again and I think it is final now.
Changes:

- I only defined one wkbLiteral now and allow for it to use the types which are defined in ExtendedWKB (the types that define 3D geometries)
- I recommend not adding an SRID into the literal. WKB does not define it, EKWB technically allows it, but I wrote in the standard that it will be ignored. The SRID should be defined as a URI like in WKT or - in the future - maybe in the knowledge graph itself
- I fixed the comments of Matt
With that this pull request is again ready for review and once approved we can merge it in wherever we maintain the sepc for 1.2 onwards

**Comment by jabhay** [comment]
We discussed the label for WKB and decided it would be better to changed it to EWKB to ensure that people were aware that it included EWKB and WKB as a proper subset of EWKB.

- [ ] Change to ewkbLiteral

**Comment by situx** [comment]
All done. geo:wkbLiteral is renamed to geo:ewkbLiteral

**Review by mperry455 — APPROVED** [review]

**Review by nicholascar — APPROVED** [review]

**Pull request merged**
